qiskit/transpiler/routing/general.py

```diff
@@ -0,0 +1,216 @@
+# -*- coding: utf-8 -*-
+
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2017, 2019.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+# Copyright 2019 Andrew M. Childs, Eddie Schoute, Cem M. Unsal
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Permutation algorithms for general graphs."""
+
+import logging
+import random
+from typing import TypeVar, Iterator, Mapping, Generic, MutableMapping, MutableSet, List, \
+    Iterable, Optional
+
+import networkx as nx
+
+from qiskit.transpiler.routing import Swap
+
+_V = TypeVar('_V')
+_T = TypeVar('_T')
+
+logger = logging.getLogger(__name__)
+
+
+class ApproximateTokenSwapper(Generic[_V]):
```

Comment:
can you make this a separate pass? it is useful outside of this PR as well. When the current mapper runs, it sometimes ends up with a final layout (placement) that is different from the initial layout. We often want to undo this by doing a permutation to go back to the initial layout (e.g. because we want to preserve statevector because we are compiling only a subcircuit). So this would be a token swapping from final_layout to initial_layout.

Reply:
It could be. I haven't made the other routing functions passes because of the issues we had identified in using them as passes. Would this work because it is only called once?

Reply:
@eddieschoute I can't see a problem in making this a pass and making it available to use in other pass managers. Am i missing something?

Reply:
I agree this swapper has a very good use case.
> We often want to undo this by doing a permutation to go back to the initial layout 

I think we can make such a pass by using your ApproximateTokenSwapper with introducing new helper functions something like:
```
def construct_partial_mapping(from: Layout, to: Layout) -> Mapping:
def CouplingMap.to_nx_graph() -> nx.Graph: 
```
(I'm not sure I understand the spec of Mapping correctly, though)
The pass would be something like:
```
class ApproxTokenTailSwap(TransformationPass):  # we'd have better class name...
    def __init__(self, coupling: CouplingMap, to: Layout = None) -> None:
        if to is None:
            to = self.property_set['layout']  # initial layout

    def run(self, mapped_dag: DAGCircuit) -> DAGCircuit:
       # this pass accepts mapped dag whereas other swap passes accept not-yet-mapped dag
        from = self.property_set['final_layout']
        target = construct_partial_mapping(from, self.to)
        swaps = ApproximateTokenSwapper(coupling.to_nx_graph()).map(target)
        # add the swaps to the end of dag
        return new_dag
```
Does this make sense to you?  @eddieschoute

I know, to do this, we need to change all swap passes to set 'final_layout' in `property_set` (or return final_layout as well as dag). So we need another discussion on it.

Reply:
Is it necessary to specialize to this use case? Can't we just reuse the TokenSwapper (maybe rewritten as a Pass) and give it the required input arguments through the property set? I think it should be possible to give it the current layout (which is the final layout in this special case) and the intended layout (which is the initial layout in this special case) and it'll compute how to get there.